metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -172,14 +173,25 @@ private FeatureControlManager(
     ControllerResult<Map<String, ApiError>> updateFeatures(
         Map<String, Short> updates,
         Map<String, FeatureUpdate.UpgradeType> upgradeTypes,
-        boolean validateOnly
+        boolean validateOnly,
+        short requestVersion
     ) {
         TreeMap<String, ApiError> results = new TreeMap<>();
         List<ApiMessageAndVersion> records =
                 BoundedList.newArrayBacked(MAX_RECORDS_PER_USER_OP);
+
+        Map<String, Short> proposedUpdatedVersions = new HashMap<>();
+        finalizedVersions.forEach(proposedUpdatedVersions::put);
+        proposedUpdatedVersions.put(MetadataVersion.FEATURE_NAME, metadataVersion.get().featureLevel());
+        updates.forEach(proposedUpdatedVersions::put);
+
         for (Entry<String, Short> entry : updates.entrySet()) {
-            results.put(entry.getKey(), updateFeature(entry.getKey(), entry.getValue(),
-                upgradeTypes.getOrDefault(entry.getKey(), FeatureUpdate.UpgradeType.UPGRADE), records));
+            ApiError error = updateFeature(entry.getKey(), entry.getValue(),
+                upgradeTypes.getOrDefault(entry.getKey(), FeatureUpdate.UpgradeType.UPGRADE), records, proposedUpdatedVersions);
+            if (requestVersion > 1 && !error.error().equals(Errors.NONE)) {
```

Comment:
Thinking about a bit more. Even for V0/V1 request, it's probably better to enforce the dependency check since it's more correct. Perhaps it's better to just fail fast for all versions?

Reply:
Is it a breaking change to only return partial results though?

Reply:
In 3.9, there is only one feature that one could set. So that behavior won't be changing even if we fail fast?

Reply:
Hmm -- 3.9 has kraft and metadata version right?

I'm also wondering about the case where someone is using an old client on a new server (maybe not common) and there could be many features

Reply:
> Hmm -- 3.9 has kraft and metadata version right?